Dockerize app

gm7y8 opened the pull request:
## Summary of change

Run all the services using docker compose 

## Related issue

https://github.com/knjk04/book-project/issues/260

## Pull request checklist

Please ensure you have done the following:

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request

- [x] Created a branch that has a descriptive name (what your branch is for a in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

If in doubt, get in touch with us via our [Gitter room](https://gitter.im/book-project-community/community)

ChFlick commented:
Hi, I'd suggest to separate your docker-compose file from the currently existing docker-compose file, e.g. in a `docker-compose.production.yml` along with `application-prod.properties`.

I think that's better for the separation of the dev/prod environments and you don't have to change the application.properties for local development every time you pull new changes from git.

gm7y8 commented:
@ChFlick  as next enhancement I was planning to add an environment variable in the existing docker-compose file so it can be used for both. base on tags it detects the type of environment. As I explored this application-prod.properties is not possible as application.properties is standard file used by spring boot.

gm7y8 commented:
@ChFlick we don't need two separate docker-compose.. just property files should work.

gm7y8 commented:
@ChFlick @knjk04  can someone approve and merge this PR?

knjk04 commented:
@gm7y8 Thanks for following up. We haven't forgotten about this, so we'll get back to you shortly.

ChFlick commented on src/main/java/com/karankumar/bookproject/BookProjectApplication.java:
```diff
@@ -18,9 +18,10 @@
 
 @SpringBootApplication(exclude = ErrorMvcAutoConfiguration.class)
-public class BookProjectApplication {
+public class BookProjectApplication extends SpringBootServletInitializer {
```
Why have you added this?

gm7y8 replied:
it helps us to deploy in war or jar file instead of running main class as earlier. Here is some info about it https://www.baeldung.com/spring-boot-servlet-initializer

ChFlick commented on docker-compose.yml:
```diff
@@ -1,23 +1,37 @@
-  db:
+  mysql:
     image: mysql:latest
+    hostname: mysql
```
```suggestion
```
The hostname is not needed - it is equal to the service name by default

gm7y8 replied:
yes, i agree. but just mentioned so it would be clear, in future if some one wants to customize it will easy. we should make this parameterized so that it can be picked up from environment variables.

ChFlick commented on README.md:
```diff
@@ -46,15 +46,26 @@ Prerequisites:
+    - Password: `password`
+
+### Start using docker-compose in production mode
+1. Clone the repository
```
Could you add a hint that one needs to build the application with the production flag before one can start the application?
`mvn package -Pproduction`

gm7y8 replied:
sure!

ChFlick commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM openjdk:11
+WORKDIR /
+ADD target/*.jar app.jar
```
```suggestion
COPY target/*.jar app.jar
```
[Best practice](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy)

gm7y8 replied:
done

ChFlick commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+RUN useradd -m myuser
+USER myuser
+EXPOSE 8080
+CMD java -jar -Dspring.profiles.active=prod  app.jar
```
```suggestion
CMD ["java", "-jar", "-Dspring.profiles.active=prod", "app.jar"]
```
[Best practice](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#cmd)

gm7y8 replied:
done

ChFlick commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM openjdk:11
```
Use openjdk:11-jre-slim for a smaller image
```suggestion
FROM openjdk:11-jre-slim
```

gm7y8 replied:
done.

gm7y8 commented:
@ChFlick  review comments are taken care of.

ChFlick reviewed (approved):
Looks good, thanks for that :)
Could you just add a dockerignore?

gm7y8 commented:
@ChFlick which files do you want to add in .dockerignore?

gm7y8 commented:
@knjk04 I see build failing some unit test failure in the Travis is someone actively looking into it?

knjk04 commented:
@gm7y8 I fixed some failing tests this morning. Once you've added the dockerignore, could you rebase? If you're still seeing failing tests, could you then let us know which tests are failing?

gm7y8 commented:
> @gm7y8 Thanks for pointing this out. I fixed some failing tests this morning. Once you've added the dockerignore, could you rebase? If you're still seeing failing tests, could you then let us know which tests are failing?

build passed now.. i don't have any rules to add to .dockerignore as we are just adding jar file to docker image build.. probably we can add empty file for now.

ChFlick commented:
> build passed now.. i don't have any rules to add to .dockerignore as we are just adding jar file to docker image build.. probably we can add empty file for now.

As we're only adding the jar file to docker doesn't it mean that we can dockerignore everything except for the jar?
I'm thinking of something like
```
*
! target/*.jar
```

gm7y8 commented:
@ChFlick  Above rules in docker Ignore may not work.. so i am leaving it empty. As we don't have any other modules involved in build context it should not cause any issue. I have tweaked the docker image copy command to avoid other files to be copied during image build..

ChFlick commented:
Weird that it did not work for you.
I just tried it locally and it worked just fine using:
```
*
!target/book-project*.jar
```

Nevertheless you're right, it can also be added later as enhancement :smile: 
So I'd suggest @knjk04 to merge the PR, I'm not allowed to :shrug:

gm7y8 commented:
> Weird that it did not work for you.
> I just tried it locally and it worked just fine using:
> 
> ```
> *
> !target/book-project*.jar
> ```
> 
> Nevertheless you're right, it can also be added later as enhancement 😄
> So I'd suggest @knjk04 to merge the PR, I'm not allowed to 🤷

Thanks, probably it might be the space causing that failure. :) let me commit this

gm7y8 commented:
@ChFlick  Thanks for taking the time to test it. @knjk04 can you merge this?

knjk04 commented:
Before I merge this in, I'll double-check this PR by running it locally. 

In the meantime, could you fix the failing build?

gm7y8 commented:
> Before I merge this in, I'll double-check this PR by running it locally.
> 
> In the meantime, could you fix the failing build?

Build failure is not related to this PR.. its due oracle jdk related

noring license option: BCL -- using GPLv2+CE by default

install-jdk.sh 2020-06-02

The command "~/bin/install-jdk.sh --target "/home/travis/oraclejdk11" --workspace "/home/travis/.cache/install-jdk" --feature "11" --license "BCL"" failed and exited with 56 during .

gm7y8 commented:
@knjk04 Can you create Task to backlog to create Integration test to verify if the docker compose and build is all fine. so that we configure for every run. This ensure sanity.

knjk04 commented:
> 
> 
> > Before I merge this in, I'll double-check this PR by running it locally.
> > In the meantime, could you fix the failing build?
> 
> Build failure is not related to this PR.. its due oracle jdk related
> 
> noring license option: BCL -- using GPLv2+CE by default
> 
> install-jdk.sh 2020-06-02
> 
> The command "~/bin/install-jdk.sh --target "/home/travis/oraclejdk11" --workspace "/home/travis/.cache/install-jdk" --feature "11" --license "BCL"" failed and exited with 56 during .

Sorry, this isn't clear to me. Are you saying that the build fails because of a license issue? 

Regardless, we still want the build to pass to prevent subsequent future builds to also pass. Is there something that you could to fix the failing build?

knjk04 commented:
> 
> 
> @knjk04 Can you create Task to backlog to create Integration test to verify if the docker compose and build is all fine. so that we configure for every run. This ensure sanity.

Sorry, could you create the issue, instead? I am not clear on what specifically needs to be tested

gm7y8 commented:
> > @knjk04 Can you create Task to backlog to create Integration test to verify if the docker compose and build is all fine. so that we configure for every run. This ensure sanity.
> 
> Sorry, could you create the issue, instead? I am not clear on what specifically needs to be tested

sure!

gm7y8 commented:
/retest

gm7y8 commented:
> > > Before I merge this in, I'll double-check this PR by running it locally.
> > > In the meantime, could you fix the failing build?
> > 
> > 
> > Build failure is not related to this PR.. its due oracle jdk related
> > noring license option: BCL -- using GPLv2+CE by default
> > install-jdk.sh 2020-06-02
> > The command "~/bin/install-jdk.sh --target "/home/travis/oraclejdk11" --workspace "/home/travis/.cache/install-jdk" --feature "11" --license "BCL"" failed and exited with 56 during .
> 
> Sorry, this isn't clear to me. Are you saying that the build fails because of a license issue?
> 
> Regardless, we still want the build to pass to prevent subsequent future builds to also pass. Is there something that you could to fix the failing build?
> 
> Update: I've seen this https://travis-ci.community/t/install-of-oracle-jdk-8-failing/3038, but I don't know if this also affects JDK 11. Besides, it's probably better to not use `trusty` since that's now EOL. Removing Oracle JDK 11 is an option, but it would be preferable if there was another solution that allows us to keep it.

Build failed due Travis environment issue. nothing much i can do there. just retriggered by adding dummy commit.

gm7y8 commented:
@knjk04 I see build has passed now.

knjk04 commented:
Perfect. I'll run this locally when I get a chance to confirm this works. I'll get back to you shortly.

knjk04 commented:
Unfortunately, this isn't working for me on Windows:. I get `ERROR: Service 'bookapp' failed to build: COPY failed: no source files were specified` when I run `docker-compose up --build`

I'll try it on Linux (Arch & Fedora) shortly.

ChFlick commented:
> Unfortunately, this isn't working for me on Windows:. I get `ERROR: Service 'bookapp' failed to build: COPY failed: no source files were specified` when I run `docker-compose up --build`
> 
> @ChFlick Did this run locally for you (both the only MySQL setup and running the prod version)?
> 
> I'll try it on Linux (Arch & Fedora) shortly.

You probably did not build the application before building docker, so there is no jar in the `target` folder.
Try running `mvn package -DskipTests -Pproduction` before.

gm7y8 commented:
> > Unfortunately, this isn't working for me on Windows:. I get `ERROR: Service 'bookapp' failed to build: COPY failed: no source files were specified` when I run `docker-compose up --build`
> > @ChFlick Did this run locally for you (both the only MySQL setup and running the prod version)?
> > I'll try it on Linux (Arch & Fedora) shortly.
> 
> You probably did not build the application before building docker, so there is no jar in the `target` folder.
> Try running `mvn package -DskipTests -Pproduction` before.

You need to try all the steps in ReadMe.. else it will not work. Not sure if build will be supported in docker.. i have tested in Linux machine only..

knjk04 commented:
I ran `mvn package -DskipTests -Pproduction`. I'll look into why it didn't work later

gm7y8 commented:
@knjk04 build may not work in window its basically due to how paths are represent  ``COPY //target//*.jar app.jar`` . As per my understanding windows is not ideal environment to build containers.. probably you can try option one to just run sql and run app locally.  Other is option is build image in *nix environment and pull image layers to local and run it this should work. Build may not work on windows.

knjk04 commented:
This works fine for me on Linux, but I get some test failures with Windows (`DateUtilsTest` and `AuthorTest`). I'm looking into this

knjk04 commented:
Update: 

I commented out the failing tests. The app runs, but I see the following:

![image](https://user-images.githubusercontent.com/11173328/91661750-d292ef80-ead5-11ea-9c16-f76c55eafdac.png)

See https://gist.github.com/knjk04/53cb1bfa87ba81fbba10e1a9c126ac88 for more details.

I've followed the suggestions (removed `pnpm-lock.yaml` and `node-modules/`, installed pnpm, ran `pnpm install`).

`mvn clean package` fixed it for in the end. I'll add the steps to the README or troubleshooting wiki page.

I'm getting a null pointer exception with the call to `setBean()` in `setBook()` in `BookForm`. I've fixed this recently, so you could you please rebase? I would like to test this PR again afterwards to double-check that currently works still works

knjk04 commented:
For some reason, Vaadin reload doesn't work anymore for me

gm7y8 commented:
> This works fine for me on Linux, but I get some test failures with Windows (`DateUtilsTest` and `AuthorTest`). The strange thing is that the tests pass in my IDE (IntelliJ), but fail with `mvn clean install`. I'm looking into this.
> 
> I've also had the same error as you @gm78y. With our current Docker setup, Windows users can use MySQL from within Docker, so it would be good to keep that functionality rather than having to make them run MySQL themselves
with this PR also you should be able to run it in windows, you should try local setup steps that should work, where you are just bringing up only the mysql container.

gm7y8 commented:
> Update:
> 
> I commented out the failing tests. The app runs, but I see the following:
> 
> ![image](https://user-images.githubusercontent.com/11173328/91661750-d292ef80-ead5-11ea-9c16-f76c55eafdac.png)
> 
> See https://gist.github.com/knjk04/53cb1bfa87ba81fbba10e1a9c126ac88 for more details.
> 
> I've followed the suggestions (removed `pnpm-lock.yaml` and `node-modules/`, installed pnpm, ran `pnpm install`).
> 
> `mvn clean package` fixed it for in the end. I'll add the steps to the README or troubleshooting wiki page.
> 
> I'm getting a null pointer exception with the call to `setBean()` in `setBook()` in `BookForm`. I've fixed this recently, so you could you please rebase? I would like to test this PR again afterwards to double-check that currently works still works

This is related to NPM local environment issue.. i have seen it once.. you should try mvn clean package should take care of it.. other way it try running production mode where all the dependencies are packaged as part of jar file.

gm7y8 commented:
@knjk04  Just suggestion, I see doing rebase will introduce lot of merge commits (already there are two merge commits). Typically when we merge git will automatically handles it as long as there are no conflicts. do you want cherry-pick this commit locally and test it?  Hopefully this commit should not even cause any issue to Testcase as this change involve any of the app functionality.  Let me know your thoughts

knjk04 commented:
> This is related to NPM local environment issue.. i have seen it once.. you should try mvn clean package should take care of it.. other way it try running production mode where all the dependencies are packaged as part of jar file.

So I had already fixed this (see my earlier earlier comment -- "mvn clean package fixed it for me in the end."). I just documented it here to remind myself to add instructions to the README :)

> @knjk04 Just suggestion, I see doing re-base will introduce merge commits (already there are two merge commits) which is not good for git history. Typically when we merge git will automatically handles it as long as there are no conflicts. do you want cherry-pick this commit locally and test it? Hopefully this commit should not even cause any issue to Testcase as this change involve any of the app functionality. Let me know your thoughts

That's fine. I can test it locally

gm7y8 commented:
> > This is related to NPM local environment issue.. i have seen it once.. you should try mvn clean package should take care of it.. other way it try running production mode where all the dependencies are packaged as part of jar file.
> 
> So I had already fixed this (see my earlier earlier comment -- "mvn clean package fixed it for me in the end."). I just documented it here to remind myself to add instructions to the README :)
> 
> > @knjk04 Just suggestion, I see doing re-base will introduce merge commits (already there are two merge commits) which is not good for git history. Typically when we merge git will automatically handles it as long as there are no conflicts. do you want cherry-pick this commit locally and test it? Hopefully this commit should not even cause any issue to Testcase as this change involve any of the app functionality. Let me know your thoughts
> 
> Rebasing shouldn't introduce merge commits. At this stage in the pull request, I think it's fine to rebase as I don't expect anyone to have any of the old commits that will get rewritten.
> 
> While I could test this locally, I'd prefer if you did a rebase in the interest of fairness; I'd rather not say it's OK to not rebase for one person and then require everyone else to rebase (see our contributing document). We want the commits to cleanly apply on top of master, which is why we ask everyone to rebase.
> 
> On the issue of Vaadin reload, do you have any thoughts as to why it doesn't work anymore?

I just rebased.. I'm not sure about vaadin reload.. did u recently upgraded the vaadin version? why don't u raise case with vaadin community?

gm7y8 commented:
Also created #289 and #290  as future enhancements.. can you assign it to me?

knjk04 commented:
> I just rebased.. I'm not sure about vaadin reload.. did u recently upgraded the vaadin version? why don't u raise case with vaadin community?

I did recently upgrade Vaadin, but I tested your PR before. 

That's fine, I was just curious if you knew how to fix it. It's not a deal-breaker; this PR can still go ahead and be merged :)

knjk04 commented:
> Also created #289 and #290 as future enhancements.. can you assign it to me?

Sure

knjk04 reviewed (approved):
Thanks a lot for this! This is really helpful! I appreciate there was a lot of to-and-fros, but we got there in the end :)
